grudge/trace_pair.py

```diff
@@ -272,61 +274,122 @@ def connected_ranks(dcoll: DiscretizationCollection):
     return get_connected_partitions(dcoll._volume_discr.mesh)
 
 
+def flatten_to_numpy_from_container(actx, ary):
+    """
+    """
+    if is_array_container(ary):
+        containerized_data_template = {}
+        flat_data = []
+        for key, value in serialize_container(ary):
+            # FIXME: Need to special case nested object arrays
+            if isinstance(value, np.ndarray) and value.dtype == "O":
+                numpy_values = flatten_to_numpy(actx, value)
+                flat_data.append(np.concatenate(numpy_values))
+            else:
+                numpy_values = actx.to_numpy(value)
+                flat_data.append(numpy_values)
+            containerized_data_template[key] = numpy_values
+        return np.concatenate(flat_data), containerized_data_template
+    else:
+        # No data template for arys with only a single state
+        # e.g. a single DOFArray
+        return flatten_to_numpy(actx, ary), None
```

Comment:
@inducer @alexfikl I think this function is fairly straightforward --- it's very similar to just regular `flatten_to_numpy` with the addition that it flattens numpy object arrays of numpy arrays. This special casing is ugly, but it seems to be rampant in `meshmode` anyways, so it's a known issue.

The reason this function _also_ returns something called a `containerized_data_template` is that this is used to help unflatten and recontainerize.

Reply:
Fixed a subtle bug in https://github.com/inducer/grudge/pull/154/commits/44968709360df8ed37edea973499ce257d602461